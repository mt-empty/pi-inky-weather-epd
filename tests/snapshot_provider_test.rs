//! Provider-specific snapshot tests
//!
//! These tests verify that the complete dashboard generation pipeline produces
//! consistent output for each weather provider (BOM and Open-Meteo).
//!
//! ## How These Tests Work
//!
//! 1. **Fixed Time**: Use FixedClock to ensure deterministic "current hour"
//! 2. **Fixture Data**: Load pre-defined weather data (no API calls)
//! 3. **Full Pipeline**: Run complete dashboard generation
//! 4. **Snapshot SVG**: Capture and compare the full SVG output
//!
//! ## Running These Tests
//!
//! **Default (Open-Meteo)**:
//! ```bash
//! TEST_MODE=config/test cargo test --test snapshot_provider_test
//! ```
//!
//! **BOM Provider** (requires override):
//! ```bash
//! TEST_MODE=config/test APP_API__PROVIDER=bom cargo test --test snapshot_provider_test snapshot_bom_dashboard -- --ignored
//! ```
//!
//! **Important**: Set `TEST_MODE=config/test` to load test configuration.
//! This loads `config/test.toml` which sets:
//! - `provider = "open_meteo"` (default test provider)
//! - `disable_weather_api_requests = true` (use fixtures, no real API calls)
//! - `weather_data_cache_path = "tests/fixtures/"` (load from test fixtures)
//!
//! ## Reviewing Snapshots
//!
//! On first run or after intentional changes:
//! ```bash
//! TEST_MODE=config/test cargo test --test snapshot_provider_test
//! cargo insta review  # Review and accept/reject changes
//! ```

use pi_inky_weather_epd::{clock::FixedClock, generate_weather_dashboard_with_clock, CONFIG};
use std::fs;

/// Test Open-Meteo provider dashboard generation with fixed time and fixtures
///
/// **Fixed Time**: Oct 10, 2025, 1:00 AM UTC = Oct 10, 2025, 12:00 PM Melbourne (AEDT)
///
/// **Fixture Files Used**:
/// - `tests/fixtures/open_meteo_forecast.json` - Combined hourly + daily data
///
/// **What This Tests**:
/// - Open-Meteo API response parsing (parallel array structure)
/// - Float temperature handling
/// - Global provider (lat/lon instead of geohash)
/// - GMT timezone conversion to Melbourne (AEDT)
/// - Dashboard rendering with Open-Meteo data structure
///
/// **Snapshot Captures**:
/// - Complete SVG structure (template + rendered data)
/// - Current weather display (temp, icon, feels like, wind, humidity, UV)
/// - Hourly forecast graph (temperature, rain)
/// - 7-day forecast cards (only 3 days in fixture)
/// - Sunrise/sunset times
///
/// **NOTE**: This is the default test provider (config/test.toml sets provider = "open_meteo").
/// No special environment variables needed.
#[test]
fn snapshot_open_meteo_dashboard() {
    // Verify test configuration is loaded
    assert!(
        CONFIG.debugging.disable_weather_api_requests,
        "Test requires TEST_MODE=config/test environment variable"
    );
    assert_eq!(
        CONFIG.misc.weather_data_cache_path.to_str().unwrap(),
        "tests/fixtures/",
        "Test config should use fixtures path"
    );
    assert_eq!(
        format!("{}", CONFIG.api.provider).to_lowercase(),
        "openmeteo",
        "Test config should use Open-Meteo provider"
    );

    // Fixed time: Oct 10, 2025, 1:00 AM UTC
    // In Melbourne AEDT (UTC+11): Oct 10, 2025, 12:00 PM (noon)
    // This should match mid-point of hourly forecast data
    let clock =
        FixedClock::from_rfc3339("2025-10-10T01:00:00Z").expect("Failed to create fixed clock");

    // Verify test configuration is loaded
    assert!(
        CONFIG.debugging.disable_weather_api_requests,
        "Test config should have disable_weather_api_requests = true"
    );

    // Temporarily override provider to Open-Meteo
    // Note: This requires the config to be reloadable or we need a different approach
    // For now, we'll test with whatever provider is set in test.toml

    // Generate the dashboard with Open-Meteo provider
    let result = generate_weather_dashboard_with_clock(&clock);
    assert!(
        result.is_ok(),
        "Failed to generate Open-Meteo dashboard: {:?}",
        result.err()
    );

    // Read the generated SVG
    let svg_content = fs::read_to_string(CONFIG.misc.generated_svg_name.clone())
        .expect("Failed to read generated SVG file");

    // Verify SVG is not empty
    assert!(!svg_content.is_empty(), "Generated SVG should not be empty");
    assert!(
        svg_content.contains("<svg"),
        "Generated file should be valid SVG"
    );

    // Snapshot the SVG structure
    // This captures the complete dashboard including:
    // - Weather data from Open-Meteo fixtures
    // - Rendered at fixed time (12:00 PM Melbourne)
    // - All graph paths, labels, icons
    insta::assert_snapshot!("open_meteo_dashboard", svg_content);
}

/// Test BOM provider dashboard generation with fixed time and fixtures
///
/// **Prerequisites**: Run with `TEST_MODE=config/test` environment variable
///
/// **Fixed Time**: Oct 9, 2025, 10:00 PM UTC = Oct 10, 2025, 9:00 AM Melbourne (AEDT)
///
/// **Fixture Files Used**:
/// - `tests/fixtures/bom_hourly_forecast.json` - 4 hours of data
/// - `tests/fixtures/bom_daily_forecast.json` - 3 days of data
///
/// **What This Tests**:
/// - BOM-specific API response parsing
/// - Integer temperature handling
/// - Knots to km/h wind conversion
/// - Australian timezone (AEDT) display
/// - Dashboard rendering with BOM data structure
///
/// **Snapshot Captures**:
/// - Complete SVG structure (template + rendered data)
/// - Current weather display (temp, icon, feels like, wind, humidity, UV)
/// - Hourly forecast graph (temperature, rain)
/// - 7-day forecast cards
/// - Sunrise/sunset times
///
/// **NOTE**: This test requires APP_API__PROVIDER environment variable override:
/// ```bash
/// TEST_MODE=config/test APP_API__PROVIDER=bom cargo test --test snapshot_provider_test snapshot_bom_dashboard -- --ignored
/// cargo insta review
/// ```
/// Note: Double underscore __ is used as separator for nested config keys (api.provider -> API__PROVIDER)
#[test]
#[ignore] // Run separately with APP_API__PROVIDER=bom (CONFIG initializes once per test binary)
fn snapshot_bom_dashboard() {
    // Verify test configuration is loaded (fail fast if TEST_MODE not set)
    assert!(
        CONFIG.debugging.disable_weather_api_requests,
        "Test requires TEST_MODE=config/test environment variable"
    );
    assert_eq!(
        CONFIG.misc.weather_data_cache_path.to_str().unwrap(),
        "tests/fixtures/",
        "Test config should use fixtures path"
    );
    assert_eq!(
        format!("{}", CONFIG.api.provider).to_lowercase(),
        "bom",
        "Test should use BOM provider (set APP_API__PROVIDER=bom)"
    );

    // Fixed time: Oct 9, 2025, 10:00 PM UTC
    // In Melbourne AEDT (UTC+11): Oct 10, 2025, 9:00 AM
    // This matches the first hourly forecast data point in the fixture
    let clock =
        FixedClock::from_rfc3339("2025-10-09T22:00:00Z").expect("Failed to create fixed clock");

    // Generate the dashboard with BOM provider
    let result = generate_weather_dashboard_with_clock(&clock);
    assert!(
        result.is_ok(),
        "Failed to generate BOM dashboard: {:?}",
        result.err()
    );

    // Read the generated SVG
    let svg_content = fs::read_to_string(CONFIG.misc.generated_svg_name.clone())
        .expect("Failed to read generated SVG file");

    // Verify SVG is not empty
    assert!(!svg_content.is_empty(), "Generated SVG should not be empty");
    assert!(
        svg_content.contains("<svg"),
        "Generated file should be valid SVG"
    );

    // Snapshot the SVG structure
    // This captures the complete dashboard including:
    // - Weather data from BOM fixtures
    // - Rendered at fixed time (9:00 AM Melbourne)
    // - All graph paths, labels, icons
    insta::assert_snapshot!("bom_dashboard", svg_content);
}
