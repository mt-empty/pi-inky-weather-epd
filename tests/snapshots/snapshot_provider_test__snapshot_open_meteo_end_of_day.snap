---
source: tests/snapshot_provider_test.rs
expression: svg_content
---
<svg width="800" height="480" font-family="Roboto, sans-serif" xmlns="http://www.w3.org/2000/svg">

    <!-- tspan doesn't work well with the degree symbol ° for some fonts!-->
    <!-- Avoid using tspan with text-anchor, `dx` or `dy`, resvg doesn't handle it properly -->
    <!-- see https://github.com/linebender/resvg/issues/583 -->

    <defs>
        <linearGradient id="fillGradient" x1="0%" y1="100%" x2="0%" y2="0%">
            <stop offset="0%" stop-color="blue" stop-opacity="0.5" />
            <stop offset="100%" stop-color="blue" stop-opacity="0.1" />
        </linearGradient>
        <pattern id="dots" patternUnits="userSpaceOnUse" width="10" height="10">
            <circle cx="5" cy="5" r="2" fill="blue" />
        </pattern>
        <pattern id="grid" patternUnits="userSpaceOnUse" width="10" height="10">
            <path d="M 10 0 L 0 0 0 10" fill="none" stroke="blue" stroke-width="1" />
        </pattern>
        <pattern id="diagonal-lines" patternUnits="userSpaceOnUse" width="10" height="10">
            <path d="M 0 10 L 10 0" stroke="blue" stroke-width="2" />
        </pattern>
        <pattern id="crosshatch" patternUnits="userSpaceOnUse" width="10" height="10">
            <path d="M 0 0 L 10 10 M 10 0 L 0 10" stroke="blue" stroke-width="1" />
        </pattern>
    </defs>

    <rect width="100%" height="100%" fill="white" />

    <!-- Current Day Information -->
    <!-- Due to resvg bug, the position is intentionally off to compensate for the bug -->
    <!-- Do not modify the x position of current_hour_temp and current_hour_feels_like, see above issue at the start of the file -->
    <text x="270" y="50" font-size="35" fill="black">Sunday, 26 October </text>
    <image x="0" y="0" width="200" height="180" href="static/fill-svg-static/overcast-night.svg" />


    <!-- Current temperature and Feels Like temperature -->
    <text x="266" y="128" font-size="70" fill="red" text-anchor="middle">
        14
        <tspan baseline-shift="27" dx="-20" font-size="33">°C</tspan>
    </text>

    <text x="246" y="158" text-anchor="middle" font-size="18" fill="black">
        <tspan>Feels</tspan>
        <tspan dy="15.5" dx="-46.6">Like</tspan>
        <tspan font-family="Roboto-Regular-Dashed" dominant-baseline="middle" font-size="55" fill="green"
            dx="10" dy="-7">
            14
            <tspan baseline-shift="12" dx="-15" font-size="30">°</tspan>
        </tspan>
    </text>

    <!-- Diagnostic Message -->
    <svg x="580" y="0" width="200" height="100" text-anchor="middle" visibility="visible">
        <!-- Cascading diagnostic icons (stacked by priority) -->
        <image x="63" y="-10" width="74" height="74" href="static/fill-svg-static/code-yellow.svg"/>
        <!-- Message for highest priority error only -->
        <text x="100" y="60" width="200" font-size="12">
            Incomplete Data
        </text>
    </svg>

    <!-- Sunset/Sunrise Information -->
    <svg x="30" y="150">
        <image x="0" y="0" width="75" height="75" href="static/fill-svg-static/sunrise.svg" />
        <text x="12" y="70" text-anchor="start" font-size="20" fill="black">06:20</text>

        <image x="70" y="0" width="75" height="75" href="static/fill-svg-static/sunset.svg" />
        <text x="83" y="70" text-anchor="start" font-size="20" fill="black">19:48</text>
    </svg>

    <g transform="translate(-15, 0)">

        <!-- vertical lines with artistic variations -->
        <svg class="line">
            <path d="M95 240 C90 220, 95 320, 95 470" />
            <path d="M150 240 C155 220, 145 320, 150 470" />

            <!-- horizontal lines with artistic variations -->
            <path d="M40 260 C150 261, 250 258, 185 260" />
            <path d="M40 330 C150 325, 250 335, 185 330" />
            <path d="M40 400 C150 402, 250 399, 185 400" />
        </svg>

        <!-- row header-->
        <svg x="15" y="210" label="UV-index" text-anchor="start">
            <text x="16" y="43.5" fill="black" font-size="20">Metric
            </text>

            <text x="85" y="43.5" fill="black" font-size="20">Now
            </text>
            <text x="140" y="43.5" fill="black" font-size="20">Max
                <tspan baseline-shift="super" font-size="17" dx="-3">24h</tspan>
            </text>
        </svg>

        <!-- UV Index -->
        <svg x="20" y="260" label="UV-index" text-anchor="middle">
            <image width="75" height="75" href="static/fill-svg-static/uv-index-none.svg" />
            <text x="100" y="45.5" font-size="30" fill="black">0
            </text>
            <text x="155" y="45.5" font-size="30" fill="black"
                font-style="normal">5
            </text>
        </svg>

        <!-- Wind Information -->
        <svg x="20" y="330" label="wind" text-anchor="middle">
            <text x="42.2" y="66.2" fill="black" font-size="15.1" text-anchor="middle">
                km/h
            </text>
            <image width="75" height="75" href="static/fill-svg-static/wind.svg" />
            <text x="100" y="45.5" font-size="30" fill="black">4
            </text>
            <text x="155" y="45.5" font-size="30" fill="black"
                font-style="normal">21
            </text>
        </svg>

        <!-- Relative Humidity -->
        <svg x="20" y="400" label="relative_humidity" text-anchor="middle">
            <image width="75" height="75" href="static/fill-svg-static/humidity-plus-plus.svg" />
            <text x="100" y="45.5" font-size="30" fill="black">90
            </text>
            <text x="155" y="45.5" font-size="30" fill="black"
                font-style="normal">91
            </text>
        </svg>
    </g>


    <!-- Forecast for the next 6 days -->
    <svg x="292" y="30">
        <text x="50" y="65" text-anchor="middle" font-size="20" fill="black">Mon</text>
        <image x="12" y="50" width="75" height="100" href="static/fill-svg-static/clear-day.svg" />
        <text y="155" font-size="20" fill="black">
            <tspan x="52" text-anchor="end">13°</tspan>
            <tspan x="52" text-anchor="middle">|</tspan>
            <tspan x="52" text-anchor="start">6°</tspan>
        </text>
    </svg>

    <svg x="373" y="30">
        <text x="50" y="65" text-anchor="middle" font-size="20" fill="black">Tue</text>
        <image x="12" y="50" width="75" height="100" href="static/fill-svg-static/clear-day.svg" />
        <text y="155" font-size="20" fill="black">
            <tspan x="52" text-anchor="end">17°</tspan>
            <tspan x="52" text-anchor="middle">|</tspan>
            <tspan x="52" text-anchor="start">8°</tspan>
        </text>
    </svg>

    <svg x="454" y="30">
        <text x="50" y="65" text-anchor="middle" font-size="20" fill="black">Wed</text>
        <image x="12" y="50" width="75" height="100" href="static/fill-svg-static/clear-day.svg" />
        <text y="155" font-size="20" fill="black">
            <tspan x="52" text-anchor="end">18°</tspan>
            <tspan x="52" text-anchor="middle">|</tspan>
            <tspan x="52" text-anchor="start">10°</tspan>
        </text>
    </svg>

    <svg x="535" y="30">
        <text x="50" y="65" text-anchor="middle" font-size="20" fill="black">Thu</text>
        <image x="12" y="50" width="75" height="100" href="static/fill-svg-static/clear-day.svg" />
        <text y="155" font-size="20" fill="black">
            <tspan x="52" text-anchor="end">19°</tspan>
            <tspan x="52" text-anchor="middle">|</tspan>
            <tspan x="52" text-anchor="start">8°</tspan>
        </text>
    </svg>

    <svg x="616" y="30">
        <text x="50" y="65" text-anchor="middle" font-size="20" fill="black">Fri</text>
        <image x="12" y="50" width="75" height="100" href="static/fill-svg-static/clear-day.svg" />
        <text y="155" font-size="20" fill="black">
            <tspan x="52" text-anchor="end">25°</tspan>
            <tspan x="52" text-anchor="middle">|</tspan>
            <tspan x="52" text-anchor="start">13°</tspan>
        </text>
    </svg>

    <svg x="697" y="30">
        <text x="50" y="65" text-anchor="middle" font-size="20" fill="black">NA</text>
        <image x="12" y="50" width="75" height="100" href="static/fill-svg-static/not-available.svg" />
        <text y="155" font-size="20" fill="black">
            <tspan x="52" text-anchor="end">NA°</tspan>
            <tspan x="52" text-anchor="middle">|</tspan>
            <tspan x="52" text-anchor="start">NA°</tspan>
        </text>
    </svg>

    <!-- Graph for hourly forecast -->
    <svg x="220" y="160" width="560" height="350" viewBox="-50 0 690 300" preserveAspectRatio="xMidYMid meet"
        style="overflow: visible;">
        <path stroke="black" stroke-linejoin="round" stroke-width="2" d="M 0 300 L 600 300 M 0 295 L 0 305 M 100 295 L 100 305 M 200 295 L 200 305 M 300 295 L 300 305 M 400 295 L 400 305 M 500 295 L 500 305 M 600 295 L 600 305" fill="none" />
        <!-- X axis Guidelines -->
        <path stroke="black" stroke-linejoin="round" stroke-width="1" d="M 0 300 L 600 300 M 0 300 v -300 m 0 2 v -2 M 100 300 v -300 m 0 2 v -2 M 200 300 v -300 m 0 2 v -2 M 300 300 v -300 m 0 2 v -2 M 400 300 v -300 m 0 2 v -2 M 500 300 v -300 m 0 2 v -2" fill="none"
            stroke-dasharray="5,10" />
        <!-- X Labels -->
        <text x="0" y="320" fill="black" font-size="19" text-anchor="middle">12am</text><text x="100" y="320" fill="black" font-size="19" text-anchor="middle">4am</text><text x="200" y="320" fill="black" font-size="19" text-anchor="middle">8am</text><text x="300" y="320" fill="black" font-size="19" text-anchor="middle">12pm</text><text x="400" y="320" fill="black" font-size="19" text-anchor="middle">4pm</text><text x="500" y="320" fill="black" font-size="19" text-anchor="middle">8pm</text><text x="600" y="320" fill="black" font-size="19" text-anchor="middle">12am</text>
        <!-- Y Labels -->
        <text x="-10" y="300"  fill="black" font-size="35" text-anchor="end" dx="8" dy="4">12°</text><text x="-10" y="240.00002"  fill="black" font-size="19" text-anchor="end" dx="8" dy="4">14.2°</text><text x="-10" y="179.99997"  fill="black" font-size="19" text-anchor="end" dx="8" dy="4">16.4°</text><text x="-10" y="119.999985"  fill="black" font-size="19" text-anchor="end" dx="8" dy="4">18.5°</text><text x="-10" y="59.999985"  fill="black" font-size="19" text-anchor="end" dx="8" dy="4">20.7°</text><text x="-10" y="0"  fill="black" font-size="35" text-anchor="end" dx="8" dy="4">23°</text>
        <!-- Y right Labels -->
        <text x="610" y="300" fill="black"  font-size="19" text-anchor="start" dy="4">0%</text><text x="610" y="240" fill="black"  font-size="19" text-anchor="start" dy="4">20%</text><text x="610" y="180" fill="black"  font-size="19" text-anchor="start" dy="4">40%</text><text x="610" y="120" fill="black"  font-size="19" text-anchor="start" dy="4">60%</text><text x="610" y="60" fill="black"  font-size="19" text-anchor="start" dy="4">80%</text><text x="610" y="0" fill="black"  font-size="19" text-anchor="start" dy="4">100%</text>
        <path stroke="red" stroke-linejoin="round" transform="translate(0, 300) scale(1, -1)"
            d="M 0.0000 41.5888C 34.7826 37.8505, 43.4783 35.5140, 52.1739 33.6449C 60.8696 31.7757, 69.5652 28.9720, 78.2609 28.0374C 86.9565 27.1028, 95.6522 27.5701, 104.3478 28.0374C 113.0435 28.5047, 121.7391 29.9065, 130.4348 30.8411C 139.1304 31.7757, 147.8261 31.7757, 156.5217 33.6449C 165.2174 35.5140, 173.9130 35.0467, 182.6087 42.0561C 191.3043 49.0654, 200.0000 64.9533, 208.6956 75.7009C 217.3913 86.4486, 226.0870 99.0654, 234.7826 106.5420C 243.4783 114.0187, 252.1739 116.3551, 260.8696 120.5607C 269.5652 124.7663, 278.2609 127.5701, 286.9565 131.7757C 295.6521 135.9813, 304.3478 131.7757, 313.0435 145.7944C 321.7391 159.8131, 330.4348 194.8598, 339.1304 215.8878C 347.8261 236.9159, 356.5217 257.9439, 365.2174 271.9626C 373.9131 285.9813, 382.6087 313.0841, 391.3044 300.0000C 400.0000 286.9159, 408.6956 214.9533, 417.3913 193.4579C 426.0869 171.9626, 434.7826 175.2336, 443.4782 171.0281C 452.1739 166.8224, 460.8696 176.1683, 469.5652 168.2243C 478.2609 160.2804, 486.9565 136.4486, 495.6522 123.3645C 504.3478 110.2804, 513.0435 99.0654, 521.7391 89.7196C 530.4348 80.3738, 539.1304 73.8318, 547.8260 67.2897C 556.5217 60.7477, 565.2173 55.6075, 573.9130 50.4673C 582.6086 45.3271, 595.6522 38.7850, 600.0000 36.4486" stroke-width="3.0" fill="none" />
        <path stroke="green" stroke-linejoin="round" stroke-dasharray="5,5"
            transform="translate(0, 300) scale(1, -1)" d="M 0.0000 40.1869C 34.7826 26.6355, 43.4783 19.6262, 52.1739 16.8224C 60.8696 14.0187, 69.5652 14.9533, 78.2609 14.0187C 86.9565 13.0841, 95.6522 10.7477, 104.3478 11.2149C 113.0435 11.6822, 121.7391 15.8878, 130.4348 16.8224C 139.1304 17.7570, 147.8261 15.4205, 156.5217 16.8224C 165.2174 18.2243, 173.9130 21.4953, 182.6087 25.2336C 191.3043 28.9720, 200.0000 34.5794, 208.6956 39.2523C 217.3913 43.9252, 226.0870 47.6635, 234.7826 53.2710C 243.4783 58.8785, 252.1739 65.8878, 260.8696 72.8972C 269.5652 79.9065, 278.2609 87.8505, 286.9565 95.3271C 295.6521 102.8037, 304.3478 100.0000, 313.0435 117.7570C 321.7391 135.5140, 330.4348 177.5701, 339.1304 201.8691C 347.8261 226.1682, 356.5217 249.0654, 365.2174 263.5514C 373.9131 278.0374, 382.6087 305.6075, 391.3044 288.7851C 400.0000 271.9626, 408.6956 188.3177, 417.3913 162.6168C 426.0869 136.9159, 434.7826 136.9159, 443.4782 134.5794C 452.1739 132.2430, 460.8696 155.1402, 469.5652 148.5981C 478.2609 142.0561, 486.9565 112.6168, 495.6522 95.3271C 504.3478 78.0374, 513.0435 55.1402, 521.7391 44.8598C 530.4348 34.5794, 539.1304 38.7850, 547.8260 33.6449C 556.5217 28.5047, 565.2173 19.6262, 573.9130 14.0187C 582.6086 8.4112, 595.6522 2.3364, 600.0000 0.0000" stroke-width="3.0" fill="none" />
        <path transform="translate(0, 300) scale(1, -1)" d="M 0.0000 174.0000L 26.086956 234L 52.173912 240L 78.260864 279L 104.347824 255L 130.43478 234L 156.52173 180L 182.60869 129L 208.69565 99L 234.78261 120L 260.86957 99L 286.9565 135L 313.04346 165L 339.13043 144L 365.21738 144L 391.30435 195L 417.3913 240L 443.47824 264L 469.56522 279L 495.65216 249L 521.73914 210L 547.82605 150L 573.913 105L 600 105 L 600 0 L 0 0Z" fill="blue"
            fill-opacity="25%" />
        <defs>
            <linearGradient id="UVGradient" gradientUnits="objectBoundingBox" x1="0%" y1="0%" x2="100%" y2="0%">
                <stop offset="0.00%" stop-color="white"/><stop offset="4.35%" stop-color="white"/><stop offset="8.70%" stop-color="white"/><stop offset="13.04%" stop-color="white"/><stop offset="17.39%" stop-color="white"/><stop offset="21.74%" stop-color="white"/><stop offset="26.09%" stop-color="white"/><stop offset="30.43%" stop-color="white"/><stop offset="34.78%" stop-color="white"/><stop offset="39.13%" stop-color="white"/><stop offset="43.48%" stop-color="white"/><stop offset="47.83%" stop-color="green"/><stop offset="52.17%" stop-color="yellow"/><stop offset="56.52%" stop-color="yellow"/><stop offset="60.87%" stop-color="white"/><stop offset="65.22%" stop-color="yellow"/><stop offset="69.57%" stop-color="green"/><stop offset="73.91%" stop-color="green"/><stop offset="78.26%" stop-color="yellow"/><stop offset="82.61%" stop-color="green"/><stop offset="86.96%" stop-color="white"/><stop offset="91.30%" stop-color="white"/><stop offset="95.65%" stop-color="white"/><stop offset="100.00%" stop-color="white"/>
            </linearGradient>
        </defs>
        <rect x="0" y="-14" width="600" height="8" fill="url(#UVGradient)" />
        <path stroke="black" stroke-linejoin="round" stroke-width="2" d="M 0 300 L 600 300 M 0 295 L 0 305 M 100 295 L 100 305 M 200 295 L 200 305 M 300 295 L 300 305 M 400 295 L 400 305 M 500 295 L 500 305 M 600 295 L 600 305" fill="none" />
        <path stroke="red" stroke-linejoin="round" stroke-width="2" d="M 0 0 L 0 300 M -5 300 L 5 300 M -5 240.00002 L 5 240.00002 M -5 179.99997 L 5 179.99997 M -5 119.999985 L 5 119.999985 M -5 59.999985 L 5 59.999985 M -5 0 L 5 0" />
        <path stroke="blue" stroke-linejoin="round" stroke-width="3" d="M 600 0 L 600 300 M 595 300 L 605 300 M 595 240 L 605 240 M 595 180 L 605 180 M 595 120 L 605 120 M 595 60 L 605 60 M 595 0 L 605 0"
            fill="none" />
    </svg>
    <!-- These are needed render svg in browser/ide -->
    <style>
        @font-face {
            font-family: 'Roboto';
            src: url('static/fonts/Roboto-Regular.ttf') format('truetype');
        }

        @font-face {
            font-family: 'Roboto-Regular-Dashed';
            src: url('static/fonts/Roboto-Regular-Dashed.ttf') format('truetype');
        }

        .line {
            stroke: black;
            stroke-width: 2;
            stroke-linecap: round;
            stroke-linejoin: round;
        }
    </style>
</svg>
